src/console_op/console.go

```diff
@@ -0,0 +1,454 @@
+		return nil
+	}
+	err := l.conn.WriteMessage(websocket.TextMessage, msg)
+	if err != nil {
+		log.Printf("WEBSOCKET Writing ERROR: %v", err)
+	}
+	return err
+}
+
+func (l *IOStreamer) Write(p []byte) (n int, err error) {
+	// NOTE: the intent is to remove the input stream contents from the
+	//  output stream so they don't get printed twice. It isn't quite working
+	//  yet, but I am leaving it in place for now.
+	//l.mu.Lock()
+	//defer l.mu.Unlock()
+
+	// remove any input commands so they don't print twice in the output
+	inStr := bytes.Clone(p)
+	//inStr = l.removeInputStrings(inStr)
+
+	// Process the remaining strings
+	n = len(inStr)
+	err = l.writeMessage(inStr)
+
+	return n, err
+}
+
+func (cs ConsoleManager) validateNode(r *http.Request) (xname, podname, errMsg string, errCode int) {
+	// Find the node requested, make sure it is valid, and if there is a tenant validate authorization
+	xname = ""
+	podname = ""
+	errMsg = ""
+	errCode = http.StatusOK
+
+	// `/console-operator/interact/{nodeXname}` - pull out the node being interacted with
+	xname = chi.URLParam(r, "nodeXname")
+	if xname == "" {
+		log.Printf("There was an error reading the node xname from the request %s", r.URL.Path)
+		errMsg = fmt.Sprintf("There was an error reading the node xname from the request %s", r.URL.Path)
+		errCode = http.StatusBadRequest
+		return
+	}
+
+	// make sure this is a valid node
+	if _, ok := nodeCache[xname]; !ok {
+		log.Printf("%s is not a valid node.", xname)
+		errMsg = fmt.Sprintf("%s is not a valid node.", xname)
+		errCode = http.StatusNotFound
+		return
+	}
+
+	// find which container is monitoring this node
+	podname, err := cs.dataService.getNodePodForXname(xname)
+	if err != nil {
+		log.Printf("Node %s is not being monitored", xname)
+		errMsg = fmt.Sprintf("Node %s is not currently being monitored", xname)
+		errCode = http.StatusNotFound
+		return
+	}
+
+	// If the user is part of a tenant, check if the node is allowed
+	tenant := r.Header.Get(tenantHeaderKey)
+	if tenant != "" && !cs.isTenantAllowed(tenant, xname) {
+		log.Printf("Tenant %s is not allowed to access node %s", tenant, xname)
+		errMsg = fmt.Sprintf("Tenant %s is not allowed to access node %s", tenant, xname)
+		errCode = http.StatusForbidden
+		return
+	}
+
+	return
+}
+
+// Handle the node interaction connection.
+func (cs ConsoleManager) doInteractConsole(w http.ResponseWriter, r *http.Request) {
+	// only allow 'GET' calls
+	if r.Method != http.MethodGet {
+		w.Header().Set("Allow", "GET")
+		sendJSONError(w, http.StatusMethodNotAllowed,
+			fmt.Sprintf("(%s) Not Allowed", r.Method))
+		return
+	}
+
+	// Make sure this is a valid operation and all required information is present
+	xname, podName, errMsg, errCode := cs.validateNode(r)
+	if errCode != http.StatusOK {
+		var body = BaseResponse{
+			Msg: errMsg,
+		}
+		SendResponseJSON(w, errCode, body)
+		return
+	}
+
+	// upgrade https to secure websocket connection
```

Comment:
Because you are implementing an API server here and dealing with web sockets, you may want to ask @jwlv to do a quick code review, looking for possible memory leak areas. He had to clean up all of the HMS Golang code for such things, and is likely going to be much better at spotting those vulnerabilities than anyone else in CSM.

I see that you are doing the `defer` calls to `Close`, but I also seem to remember there being a memory leak issue if the request body is not read before the connection is closed. I don't know how that works in this case where you are feeding the request into the `Upgrade` function to get the `conn` object.

Reply:
Josh reviewed and comments are below.